mesh/mesh_operators.cpp

```diff
@@ -157,6 +157,141 @@ int ThresholdDerefiner::ApplyImpl(Mesh &mesh)
+
+   int dim = mesh.Dimension();
+   L2_FECollection l2fec(order, dim);
+   FiniteElementSpace* l2fes = NULL;
+
+   bool par = false;
+   GridFunction *gf = NULL;
+
+#ifdef MFEM_USE_MPI
+   ParMesh* pmesh = dynamic_cast<ParMesh*>(&mesh);
+   if (pmesh && pmesh->Nonconforming())
+   {
+      par = true;
+      l2fes = new ParFiniteElementSpace(pmesh, &l2fec);
+      gf = new ParGridFunction(static_cast<ParFiniteElementSpace*>(l2fes));
+   }
+#endif
+   if (!par)
+   {
+      l2fes = new FiniteElementSpace(&mesh, &l2fec);
+      gf = new GridFunction(l2fes);
+   }
+
+   // If custom integration rule has not been set,
+   // then use the default integration rule
+   if (!irs)
+   {
+      int order_quad = 2*order + 3;
+      for (int i=0; i < Geometry::NumGeom; ++i)
+      {
+         ir_default[i] = &(IntRules.Get(i, order_quad));
+      }
+      irs = ir_default;
+   }
+
+   for (int i = 0; i < max_it; i++)
+   {
+      // Compute number of elements and L2-norm of f.
+      int NE = mesh.GetNE();
+      int globalNE = 0;
+      double norm_of_coeff = 0.0;
+      if (par)
+      {
+#ifdef MFEM_USE_MPI
+         globalNE = pmesh->GetGlobalNE();
+         norm_of_coeff = ComputeGlobalLpNorm(2.0,*coeff,*pmesh,irs);
+#endif
+      }
+      else
+      {
+         globalNE = NE;
+         norm_of_coeff = ComputeLpNorm(2.0,*coeff,mesh,irs);
+      }
+
+      // Compute average L2-norm of f
+      double av_norm_of_coeff = norm_of_coeff / sqrt(globalNE);
+
+      // Compute element-wise L2-norms of (I - Π) f
+      Vector element_norms_of_fine_scale(NE);
+      gf->SetSpace(l2fes);
+      gf->ProjectCoefficient(*coeff);
+      gf->ComputeElementL2Errors(*coeff,element_norms_of_fine_scale,irs);
+
+      // Define osc_K(f) := || h ⋅ (I - Π) f ||_K and select elements
+      // for refinement based on threshold. Also record relative osc(f).
+      global_osc = 0.0;
+      mesh_refinements.SetSize(0);
+      element_oscs.Destroy();
+      element_oscs.SetSize(NE);
+      element_oscs = 0.0;
+      for (int j = 0; j < NE; j++)
+      {
+         double h = mesh.GetElementSize(j);
+         double element_osc = h * element_norms_of_fine_scale(j);
+         if ( element_osc > threshold * av_norm_of_coeff )
+         {
+            mesh_refinements.Append(j);
+         }
+         element_oscs(j) = element_osc/(norm_of_coeff + 1e-10);
+         global_osc += element_osc*element_osc;
+      }
+#ifdef MFEM_USE_MPI
+      if (par)
+      {
+         MPI_Comm comm = pmesh->GetComm();
+         MPI_Allreduce(MPI_IN_PLACE, &global_osc, 1, MPI_DOUBLE, MPI_SUM, comm);
+         MPI_Comm_rank(comm, &rank);
+      }
+#endif
+      global_osc = sqrt(global_osc)/(norm_of_coeff + 1e-10);
+
+      // Exit if the global threshold or maximum number of elements is reached.
+      if (global_osc < threshold || globalNE > max_elements)
+      {
+         if (global_osc > threshold && globalNE > max_elements && rank == 0)
```

Comment:
One small point. I would suggest adding a flag to disable this warning similar to the `print_level` used in our iterative solvers. Some people have a strong aversion to warnings and they may want to be able to disable them in some situations. For example if MFEM's `example` codes produce warnings some test runs might lead to questions e.g. "was this warning expected or should we be concerned?" .